### PR TITLE
Pinecone as Local Online Provider

### PR DESCRIPTION
**Pull request by epps**
# Description

This PR adds support for using Pinecone as a local online provider in local mode.

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: client/tests/test_online.py [hunk 1]
```diff
@@ -0,0 +1,841 @@
...
+            "config_fn": get_pinecone_config,
+            "online_store_class": PineconeOnlineStore,
+        },
+        id="Pinecone",
+    ),
+]
+
+
+@pytest.mark.parametrize("case", cases)
+def test_online_vector_stores(case):
```
Comment (epps) [inline client/tests/test_online.py:45]: I realize I forgot to add this to the Makefile. I'll do that shortly.

File: client/src/featureform/register.py [hunk 2]
```diff
@@ -1926,6 +1927,55 @@ def register_redis(
...
+    ):
+        """Register a Pinecone provider.
+        **Examples**:
+        ```
+        pinecone = ff.register_pinecone(
+            name="pinecone-quickstart",
+            project_id="2g13ek7",
+            environment="us-west4-gcp-free",
+            api_key="e4egd064-1vb6-497f-aadf-7547atbb517f"
+            description="A Pinecone project for we Featureform embeddings"
```
Comment (epps) [inline client/src/featureform/register.py:1949]: TODO: fix this typo

File: client/src/featureform/online.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            return [match["metadata"]["id"] for match in matches]
+
+    def _create_id(self, entity: str) -> str:
+        id_uuid = uuid.uuid5(uuid.NAMESPACE_DNS, entity)
+        return str(id_uuid)
+
+
+class PineconeOnlineStore(VectorStore):
+    def __init__(self, config: PineconeConfig):
+        self.indexNameTemplate = "ff-idx--{0}"
```
Comment (epps) [inline client/src/featureform/online.py:121]: Change to snake case.

File: client/src/featureform/online.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+from abc import ABC, abstractmethod
```
Comment (aolfat) [inline client/src/featureform/online.py:1]: nice

File: client/src/featureform/online.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+from typing import List, Any, Union
+
+import numpy as np
+from .enums import ScalarType
+from .resources import PineconeConfig
+import pinecone
+import uuid
+
```
Comment (aolfat) [inline client/src/featureform/online.py:9]: I'm thinking the package structure for this should be

-online_store
--online.py (interfaces to implement)
--pinecone.py
--weaviate.py
--redis.py
--other_online_store.py

main point is that as we add online stores i don't think they should all be in 1 file.

File: client/src/featureform/online.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+class ValueType(ABC):
```
Comment (aolfat) [inline client/src/featureform/online.py:11]: are there other value types?

File: client/src/featureform/online.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+class VectorType(ValueType):
+    def __init__(self, scalar_type: ScalarType, dimension: int, is_embedding: bool):
+        self.scalar_type = scalar_type
+        self.dimension = dimension
+        self.is_embedding = is_embedding
+
+    def scalar(self):
+        return self.scalar_type
+
+    def is_vector(self):
```
Comment (aolfat) [inline client/src/featureform/online.py:30]: I think the abstraction could be leveraged instead. Static method on ValueType. isVectorType(value).

File: client/src/featureform/online.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        pass
+
+    @abstractmethod
+    def delete_index(self, feature: str, variant: str):
+        pass
+
+
+class VectorStoreTable(OnlineStoreTable, ABC):
+    @abstractmethod
+    def nearest(self, feature: str, variant: str, vector: List[float], k: int):
```
Comment (aolfat) [inline client/src/featureform/online.py:74]: nit: some docstrings would be nice for a lot of these.

File: client/tests/test_online.py [hunk 9]
```diff
@@ -0,0 +1,841 @@
...
+    for _, row in embeddings.iterrows():
+        vector = np.fromstring(row["headline_vector"][1:-1], sep=", ")
+        table.set(row["publisher"], vector.tolist())
+
+    expected = embeddings.tail(1)
+    actual = table.get(expected["publisher"].values[0])
+    assert actual is not None and actual == expected["publisher"].values[0]
+
+    # Test nearest
+    query_vector = [
```
Comment (aolfat) [inline client/tests/test_online.py:68]: holy shit does this have to be this long lol

File: client/src/featureform/online.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    def scalar(self):
+        return self.scalar_type
+
+    def is_vector(self):
+        return True
+
+
+class OnlineStore(ABC):
+    @abstractmethod
+    def get_table(self, feature: str, variant: str):
```
Comment (aolfat) [inline client/src/featureform/online.py:36]: returns OnlineTable?

File: client/tests/test_online.py (hunk 9 above)
Comment (epps) [reply client/tests/test_online.py:68]: I either read it from a file and deserialize or hardcode it here 🤷‍♂️  In either case, the dimensions have to match the index.

**Pull request merged**